### PR TITLE
require spec.serviceAccountNames

### PR DESCRIPTION
**Pull request by joelddiaz**
No longer treat CredentialsRequests without spec.serviceAccountNames
filled out as valid. This creates IAM Roles that limit the Role by
audience which means any ServiceAccount in the cluster could potentially
assume the IAM Role (because they all share the same audience).

All CredentialsRequests now populate the ServiceAccountNames field, so
we can remove this transitional support.

xref: https://issues.redhat.com/browse/CCO-104

### CODE REVIEW

File: pkg/cmd/provisioning/aws/create-iam-roles.go [hunk 1]
```diff
@@ -256,7 +256,11 @@ func createRolePolicyDocument(oidcProviderARN, issuerURL, namespace string, serv
 		conditionString = fmt.Sprintf(` { "StringEquals": { "%s:sub": %s } }`, issuerURL, serviceAccountListString)
 	} else {
 		// TODO: maybe return an error once all CredentialsRequest start including ServiceAccountNames
-		conditionString = fmt.Sprintf(`{ "StringEquals": { "%s:aud": "openshift" } }`, issuerURL)
+		// We used to support leaving the list of ServiceAccounts blank in the CredentialsRequest while
+		// we transitioned all the existing CredReqs. That work is complete, and we should not
+		// create Role policies that limit the Role by audience any more.
+		// Return an error indicating that ccoctl requires the ServiceAccount list to be populated.
+		return "", fmt.Errorf("must provide ServieAccounts to bind the Role policy to")
```
Comment (akhil-rane) [inline pkg/cmd/provisioning/aws/create-iam-roles.go:263]:
```suggestion
		return "", fmt.Errorf("CredentialsRequest must provide ServieAccounts to bind the Role policy to")
```

Comment (joelddiaz) [reply pkg/cmd/provisioning/aws/create-iam-roles.go]: :+1: should be in the latest push

**Comment by joelddiaz** [comment]
/assign @lwan-wanglin @jeana-redhat @sferich888 
The TL;DR is: when implementing the first version of `ccoctl` we allowed not specifying the ServiceAccount list (as not all repos had added that info yet). Now all in-cluster operators have updated their CredentialsRequest manifests to include the ServiceAccounts. Creating IAM Roles that bind to a specific ServiceAccount instead of to the cluster-wide `audience` field is more secure, and we will now raise an error if someone tries to use `ccoctl` to create an IAM Role where the CredentialsRequest has not filled out the ServiceAccount names list.

**Comment by jeana-redhat** [comment]
/label docs-apporoved
thanks for the summary, Joel!

**Comment by jeana-redhat** [comment]
thanks for the summary, Joel!

/label docs-approved

**Comment by lwan-wanglin** [comment]
/label qe-approved

test pass:
prepare a CredentialsRequest without  .spec.serviceAccountNames, run ccoctl to create policies, it will hit below error:
`
2021/09/26 15:50:51 Failed to process IAM Roles: Failed while processing each CredentialsRequest: error while creating Role policy document for openshift-machine-api-aws: CredentialsRequest must provide ServieAccounts to bind the Role policy to
`
the test case: https://url.corp.redhat.com/bbfbefd

**Pull request merged**